File: server/matchmaker.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json, time, random

queues = {}
tickets = {}

def key(game, wager):
    return f"{game}:{wager}"

class H(BaseHTTPRequestHandler):
    def _send(self, code, obj):
        raw = json.dumps(obj).encode()
        self.send_response(code)
        self.send_header("content-type", "application/json")
        self.send_header("access-control-allow-origin", "*")
        self.send_header("access-control-allow-headers", "content-type")
        self.send_header("content-length", str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)

    def do_OPTIONS(self):
        self._send(204, {})

    def do_POST(self):
        if self.path != "/join":
            return self._send(404, {"error": "not found"})
        n = int(self.headers.get("content-length", "0"))
        body = json.loads(self.rfile.read(n) or b"{}")
        game, wager, wallet = body.get("game"), str(body.get("wager", "")), body.get("wallet")
        if not game or not wager or not wallet:
            return self._send(400, {"error": "game, wager, wallet required"})
        ticket = f"{int(time.time())}-{random.randrange(100000,999999)}"
        player = {"ticket": ticket, "wallet": wallet, "game": game, "wager": wager, "status": "waiting", "matchId": None, "opponent": None}
        tickets[ticket] = player
        k = key(game, wager)
        waiting = next((p for p in queues.get(k, []) if p["status"] == "waiting" and p["wallet"] != wallet), None)
        if waiting:
            mid = f"m-{int(time.time())}"
            waiting["status"] = "matched"; waiting["matchId"] = mid; waiting["opponent"] = wallet
            player["status"] = "matched"; player["matchId"] = mid; player["opponent"] = waiting["wallet"]
        else:
            queues.setdefault(k, []).append(player)
        self._send(200, {"ticket": ticket, "status": player["status"]})

    def do_GET(self):
        if not self.path.startswith("/status/"):
            return self._send(404, {"error": "not found"})
        ticket = self.path.split("/status/")[1]
        player = tickets.get(ticket)
        if not player:
            return self._send(404, {"error": "ticket not found"})
        self._send(200, {"status": player["status"], "matchId": player["matchId"], "opponent": player["opponent"]})

    def log_message(self, *a):
        pass

print("matchmaker on 8787")
HTTPServer(("0.0.0.0", 8787), H).serve_forever()

File: server/matchmaker.mjs
import http from "http";

const PORT = process.env.PORT || 8787;
const queues = new Map();
const tickets = new Map();

function key(game, wager) {
  return `${game}:${wager}`;
}

function send(res, code, obj) {
  res.writeHead(code, {
    "content-type": "application/json",
    "access-control-allow-origin": "*",
    "access-control-allow-headers": "content-type",
  });
  res.end(JSON.stringify(obj));
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    let b = "";
    req.on("data", (c) => (b += c));
    req.on("end", () => {
      try { resolve(b ? JSON.parse(b) : {}); }
      catch (e) { reject(e); }
    });
  });
}

const server = http.createServer(async (req, res) => {
  if (req.method === "OPTIONS") return send(res, 204, {});
  try {
    if (req.method === "POST" && req.url === "/join") {
      const { game, wager, wallet } = await readBody(req);
      if (!game || !wager || !wallet) return send(res, 400, { error: "game, wager, wallet required" });
      const ticket = `${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
      const k = key(game, wager);
      const list = queues.get(k) || [];
      const player = { ticket, wallet, game, wager, status: "waiting", matchId: null, opponent: null };
      tickets.set(ticket, player);
      const waiting = list.find((p) => p.status === "waiting" && p.wallet !== wallet);
      if (waiting) {
        const matchId = `m-${Date.now()}`;
        waiting.status = "matched";
        waiting.matchId = matchId;
        waiting.opponent = wallet;
        player.status = "matched";
        player.matchId = matchId;
        player.opponent = waiting.wallet;
      } else {
        list.push(player);
        queues.set(k, list);
      }
      return send(res, 200, { ticket, status: player.status, matchId: player.matchId });
    }
    if (req.method === "GET" && req.url.startsWith("/status/")) {
      const ticket = req.url.split("/status/")[1];
      const player = tickets.get(ticket);
      if (!player) return send(res, 404, { error: "ticket not found" });
      return send(res, 200, {
        status: player.status,
        matchId: player.matchId,
        opponent: player.opponent,
      });
    }
    send(res, 404, { error: "not found" });
  } catch (e) {
    send(res, 500, { error: String(e.message || e) });
  }
});

server.listen(PORT, () => console.log("matchmaker on " + PORT));
